packages/browser-destinations/src/destinations/heap/index.ts

```diff
@@ -0,0 +1,60 @@
+import type { Settings } from './generated-types'
+import type { BrowserDestinationDefinition } from '../../lib/browser-destinations'
+import { browserDestination } from '../../runtime/shim'
+
+import { HeapApi } from './types'
+
+declare global {
+  interface Window {
+    heap: HeapApi
+  }
+}
+
+// Switch from unknown to the partner SDK client types
+export const destination: BrowserDestinationDefinition<Settings, HeapApi> = {
+  name: 'Heap',
+  slug: 'heap',
+  mode: 'device',
+
+  settings: {
+    // Add any Segment destination settings required here
+    appId: {
+      label: 'Heap app ID',
+      description:
+        'The app ID of the environment to which you want to send data. You can find this ID on the [Projects](https://heapanalytics.com/app/manage/projects) page.',
+      type: 'string',
+      required: true
+    },
+    disableTextCapture: {
+      label: 'Global data redaction via Disabling Text Capture',
+      description:
+        'Setting to true will redact all target text on your website. For more information visit the heap [docs page](https://developers.heap.io/docs/web#global-data-redaction-via-disabling-text-capture).',
+      type: 'boolean',
+      required: false
+    },
+    secureCookie: {
+      label: 'Secure Cookie',
+      description:
+        'This option is turned off by default to accommodate websites not served over HTTPS. If your application uses HTTPS, we recommend enabling secure cookies to prevent Heap cookies from being observed by unauthorized parties. For more information visit the heap [docs page](https://developers.heap.io/docs/web#securecookie).',
+      type: 'boolean',
+      required: false
+    }
+  },
+
+  initialize: async ({ settings }, deps) => {
+    const config = {
+      disableTextCapture: settings.disableTextCapture || false,
+      secureCookie: settings.secureCookie || false
+    }
+
+    await deps.loadScript(`https://cdn.heapanalytics.com/js/heap-${settings.appId}.js`)
+    await deps.resolveWhen(() => Object.prototype.hasOwnProperty.call(window, 'heap'), 100)
```

Comment:
This line is equivalent to all of [this](https://github.com/segmentio/analytics.js-integrations/blob/4ddb836a2db5a980fb396d70d309fc1e0d13810b/integrations/heap/lib/index.js#L36-L64). [This line ](https://github.com/segmentio/analytics.js-integrations/blob/4ddb836a2db5a980fb396d70d309fc1e0d13810b/integrations/heap/lib/index.js#L43) saves the method that was called, so we can replay the actions later explained [here](https://stackoverflow.com/questions/14859058/why-does-the-segment-io-loader-script-push-method-names-args-onto-a-queue-which). We can now use `resolveWhen` because it will [wait](https://github.com/segmentio/action-destinations/blob/2788c2bf18fac723e3834f965b05e6aa96cc1117/packages/browser-destinations/src/runtime/resolve-when.ts#L13) until `window.heap` is ready then call our method.

Reply:
😲

Reply:
this is pretty slick